src/main/java/frc/robot/poseestimator/PoseEstimationMath.java

```diff
@@ -0,0 +1,166 @@
+package frc.robot.poseestimator;
+
+import edu.wpi.first.math.geometry.*;
+import frc.robot.constants.Field;
+import frc.robot.poseestimator.observations.IRobotPoseVisionObservation;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+public class PoseEstimationMath {
+
+	public static Twist2d addGyroToTwist(Twist2d twist, Rotation2d currentGyroAngle, Rotation2d lastGyroAngle) {
+		boolean hasGyroUpdated = currentGyroAngle != null;
+		if (hasGyroUpdated && lastGyroAngle != null) {
+			return updateChangeInAngle(twist, currentGyroAngle, lastGyroAngle);
+		}
+		return twist;
+	}
+
+	public static Twist2d updateChangeInAngle(Twist2d twist, Rotation2d currentGyroAngle, Rotation2d lastGyroAngle) {
+		Rotation2d rotationDifference = currentGyroAngle.minus(lastGyroAngle);
+		return new Twist2d(twist.dx, twist.dy, rotationDifference.getRadians());
+	}
+
+	public static double[] getKalmanRatio(double[] odometryStandardDeviations, double[] visionStandardDeviations) {
+		double[] combinedStandardDeviations = new double[Pose2dArrayValue.POSE_ARRAY_LENGTH];
+		for (int i = 0; i < combinedStandardDeviations.length; i++) {
+			double odometryStandardDeviation = odometryStandardDeviations[i];
+			double visionStandardDeviation = visionStandardDeviations[i];
+			combinedStandardDeviations[i] = getKalmanRatio(odometryStandardDeviation, visionStandardDeviation);
+		}
+		return combinedStandardDeviations;
+	}
+
+	public static double getKalmanRatio(double odometryStandardDeviation, double visionStandardDeviation) {
+		return odometryStandardDeviation == 0 ? 0 : odometryStandardDeviation / (odometryStandardDeviation + visionStandardDeviation);
```

Comment:
? no need to check if the odometryStandardDeviation is equal to 0. There would be a problem if both the vision and odometry stdDevs are 0, but I don't see a scenario in which that could happen